### DISCUSSION TITLE
Question about `require.context` and circular dependencies

### DISCUSSION THREAD

**Discussion by issammani** [webpack/webpack#17256] [Q&A] [answered]
### Context
I am working on some code that uses a custom import syntax not part of any standard.  I am using `require.context` to make a custom import function accessible in to my code that mimics the import synatx. I have a use case for circular dependencies and want to know if there is any way to solve this in webpack. Having a circular dependency causes this error: `Uncaught ReferenceError: can't access lexical declaration '__WEBPACK_DEFAULT_EXPORT__' before initialization`. 

I assume it makes sense since require is sync and needs the module to be resolved when it's imported. Is there a way though to do this using webpack though ?

Here is some minimal code to reproduce:

```js
// index.js
const internalRequire = (modulePath) => {
  const moduleResolver = require.context("./", false, /.mjs$/);
  return moduleResolver(modulePath);
};

const moduleA = internalRequire("./a.mjs");
const moduleB = internalRequire("./b.mjs");
```

```js
// a.js
import b from "./b.mjs";

var a = { ...b, a: "a" };
export default a;
```

```js
// b.js
import a from "./a.mjs";

var b = { ...a, b: "b" };
export default b;
```

```js 
// webpack.config.js
const path = require("path");
module.exports = {
  entry: "./src/index.mjs",
  output: {
    filename: "bundle.js",
    path: path.resolve(__dirname, "dist"),
  },
  module: {
    rules: [
      {
        test: /\.mjs$/,
        type: "javascript/auto",
      },
    ],
  },
  mode: "development",
};
```
P.S: I am fairly certain that I can use dynamic imports here because they return promises. I just want to know whether it's feasible with the current code.

Thanks 😄 !

**Comment by alexander-akait** [comment, marked as answer]
It means you have circular deps, which can't be resolved

**Reply by alexander-akait** [reply]
Also for ESM, please use `import.meta.webpackContext()`
